File: sdk/typescript/src/namespaces/workflows.ts
/**
 * Workflows Namespace API
 *
 * Provides a namespaced interface for workflow operations.
 */

import type { AragoraClient } from '../client';

/**
 * Workflows API namespace.
 *
 * Provides methods for managing automated workflows:
 * - Creating, listing, and managing workflows
 * - Executing workflows and tracking their status
 * - Working with workflow templates
 * - Managing workflow executions
 */
export class WorkflowsAPI {
  constructor(private client: AragoraClient) {}

  // ===========================================================================
  // Workflows CRUD
  // ===========================================================================

  /**
   * List all workflows.
   * @route GET /api/v1/workflows
   */
  async list(params?: { limit?: number; offset?: number }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflows', {
      params: params as Record<string, unknown>,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Create a new workflow.
   * @route POST /api/v1/workflows
   */
  async create(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get a workflow by ID.
   * @route GET /api/v1/workflows/{workflow_id}
   */
  async get(workflowId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Update a workflow.
   * @route PUT /api/v1/workflows/{workflow_id}
   */
  async update(workflowId: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request(
      'PUT',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}`,
      { body }
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Patch a workflow.
   * @route PATCH /api/v1/workflows/{workflow_id}
   */
  async patch(workflowId: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request(
      'PATCH',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}`,
      { body }
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Delete a workflow.
   * @route DELETE /api/v1/workflows/{workflow_id}
   */
  async delete(workflowId: string): Promise<void> {
    return this.client.request(
      'DELETE',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}`
    ) as Promise<void>;
  }

  /**
   * Execute a workflow.
   * @route POST /api/v1/workflows/{workflow_id}/execute
   */
  async execute(workflowId: string, inputs?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}/execute`,
      { body: inputs }
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Simulate a workflow without executing.
   * @route POST /api/v1/workflows/{workflow_id}/simulate
   */
  async simulate(workflowId: string, inputs?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}/simulate`,
      { body: inputs }
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Get workflow status.
   * @route GET /api/v1/workflows/{workflow_id}/status
   */
  async getStatus(workflowId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}/status`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Get workflow versions.
   * @route GET /api/v1/workflows/{workflow_id}/versions
   */
  async getVersions(workflowId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}/versions`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Restore a specific version of a workflow.
   * @route POST /api/v1/workflows/{workflow_id}/versions/{version}/restore
   */
  async restoreVersion(workflowId: string, version: number): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/workflows/${encodeURIComponent(workflowId)}/versions/${version}/restore`
    ) as Promise<Record<string, unknown>>;
  }

  // ===========================================================================
  // Executions
  // ===========================================================================

  /**
   * List workflow executions.
   * @route POST /api/v1/workflows/executions
   */
  async listExecutions(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/executions', {
      body: params,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get a workflow execution by ID.
   * @route GET /api/v1/workflows/executions/{execution_id}
   */
  async getExecution(executionId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflows/executions/${encodeURIComponent(executionId)}`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Delete a workflow execution.
   * @route DELETE /api/v1/workflows/executions/{execution_id}
   */
  async deleteExecution(executionId: string): Promise<void> {
    return this.client.request(
      'DELETE',
      `/api/v1/workflows/executions/${encodeURIComponent(executionId)}`
    ) as Promise<void>;
  }

  /**
   * Update executions.
   * @route PATCH /api/v1/workflows/executions
   */
  async updateExecutions(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('PATCH', '/api/v1/workflows/executions', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Replace executions.
   * @route PUT /api/v1/workflows/executions
   */
  async replaceExecutions(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('PUT', '/api/v1/workflows/executions', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Delete executions.
   * @route DELETE /api/v1/workflows/executions
   */
  async deleteExecutions(): Promise<void> {
    return this.client.request('DELETE', '/api/v1/workflows/executions') as Promise<void>;
  }

  // ===========================================================================
  // Templates
  // ===========================================================================

  /**
   * Create a workflow template.
   * @route POST /api/v1/workflows/templates
   */
  async createTemplate(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/templates', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get a workflow template by ID.
   * @route GET /api/v1/workflows/templates/{template_id}
   */
  async getTemplate(templateId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflows/templates/${encodeURIComponent(templateId)}`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Update a workflow template.
   * @route PUT /api/v1/workflows/templates/{template_id}
   */
  async updateTemplate(templateId: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request(
      'PUT',
      `/api/v1/workflows/templates/${encodeURIComponent(templateId)}`,
      { body }
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Delete a workflow template.
   * @route DELETE /api/v1/workflows/templates/{template_id}
   */
  async deleteTemplate(templateId: string): Promise<void> {
    return this.client.request(
      'DELETE',
      `/api/v1/workflows/templates/${encodeURIComponent(templateId)}`
    ) as Promise<void>;
  }

  /**
   * Patch templates.
   * @route PATCH /api/v1/workflows/templates
   */
  async patchTemplates(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('PATCH', '/api/v1/workflows/templates', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Replace templates.
   * @route PUT /api/v1/workflows/templates
   */
  async replaceTemplates(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('PUT', '/api/v1/workflows/templates', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Delete templates.
   * @route DELETE /api/v1/workflows/templates
   */
  async deleteTemplates(): Promise<void> {
    return this.client.request('DELETE', '/api/v1/workflows/templates') as Promise<void>;
  }

  // ===========================================================================
  // Workflow Templates (top-level /workflow-templates path)
  // ===========================================================================

  /**
   * List workflow templates.
   * @route GET /api/v1/workflow-templates
   */
  async listWorkflowTemplates(params?: { category?: string }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflow-templates', {
      params: params as Record<string, unknown>,
    }) as Promise<Record<string, unknown>>;
  }

  // ===========================================================================
  // Workflow Executions (top-level /workflow-executions path)
  // ===========================================================================

  /**
   * List workflow executions.
   * @route GET /api/v1/workflow-executions
   */
  async listWorkflowExecutions(params?: {
    workflow_id?: string;
    status?: string;
    limit?: number;
  }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflow-executions', {
      params: params as Record<string, unknown>,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get a workflow execution by ID.
   * @route GET /api/v1/workflow-executions/{execution_id}
   */
  async getWorkflowExecution(executionId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/workflow-executions/${encodeURIComponent(executionId)}`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Terminate a running workflow execution.
   * @route DELETE /api/v1/workflow-executions/{execution_id}
   */
  async terminateExecution(executionId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'DELETE',
      `/api/v1/workflow-executions/${encodeURIComponent(executionId)}`
    ) as Promise<Record<string, unknown>>;
  }

  // ===========================================================================
  // Approvals
  // ===========================================================================

  /**
   * List pending workflow approvals.
   * @route GET /api/v1/workflow-approvals
   */
  async listApprovals(params?: {
    workflow_id?: string;
    tenant_id?: string;
  }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflow-approvals', {
      params: params as Record<string, unknown>,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Resolve a pending workflow approval request.
   * @route POST /api/v1/workflow-approvals/{request_id}/resolve
   */
  async resolveApproval(
    requestId: string,
    body: { status?: string; notes?: string; checklist?: Record<string, unknown> }
  ): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/workflow-approvals/${encodeURIComponent(requestId)}/resolve`,
      { body }
    ) as Promise<Record<string, unknown>>;
  }

  // ===========================================================================
  // Visual Builder
  // ===========================================================================

  /**
   * Auto-layout workflow steps for visual rendering.
   * @route POST /api/v1/workflows/auto-layout
   */
  async autoLayout(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/auto-layout', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Create a workflow from a known pattern.
   * @route POST /api/v1/workflows/from-pattern
   */
  async fromPattern(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/from-pattern', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Generate a workflow from natural language input.
   * @route POST /api/v1/workflows/generate
   */
  async generate(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/generate', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * List supported workflow builder step types.
   * @route GET /api/v1/workflows/step-types
   */
  async listStepTypes(params?: { category?: string }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflows/step-types', {
      params: params as Record<string, unknown>,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Validate a workflow definition.
   * @route POST /api/v1/workflows/validate
   */
  async validate(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/workflows/validate', {
      body,
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Replay a workflow with provided inputs.
   * @route POST /api/v1/workflows/{workflow_id}/replay
   */
  async replay(workflowId: string, body?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', `/api/v1/workflows/${encodeURIComponent(workflowId)}/replay`, {
      body,
    }) as Promise<Record<string, unknown>>;
  }
}

File: sdk/python/aragora/namespaces/workflows.py
"""
Workflows Namespace API

Provides methods for creating and executing automated workflows.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class WorkflowsAPI:
    """
    Synchronous Workflows API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> execution = client.workflows.execute("workflow-123", {"input": "value"})
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(
        self,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List all workflows.

        Args:
            limit: Maximum number of workflows to return
            offset: Number of workflows to skip

        Returns:
            List of workflows
        """
        return self._client.request(
            "GET",
            "/api/v1/workflows",
            params={"limit": limit, "offset": offset},
        )

    def get(self, workflow_id: str) -> dict[str, Any]:
        """
        Get a workflow by ID.

        Args:
            workflow_id: The workflow ID

        Returns:
            Workflow details
        """
        return self._client.request("GET", f"/api/v1/workflows/{workflow_id}")

    def create(
        self,
        name: str,
        steps: list[dict[str, Any]],
        description: str | None = None,
        **kwargs,
    ) -> dict[str, Any]:
        """
        Create a new workflow.

        Args:
            name: Workflow name
            steps: List of workflow steps
            description: Optional description
            **kwargs: Additional workflow options

        Returns:
            Created workflow
        """
        data = {"name": name, "steps": steps, **kwargs}
        if description:
            data["description"] = description

        return self._client.request("POST", "/api/v1/workflows", json=data)

    def update(
        self,
        workflow_id: str,
        **updates,
    ) -> dict[str, Any]:
        """
        Update a workflow.

        Args:
            workflow_id: The workflow ID
            **updates: Fields to update

        Returns:
            Updated workflow
        """
        return self._client.request(
            "PUT",
            f"/api/v1/workflows/{workflow_id}",
            json=updates,
        )

    def delete(self, workflow_id: str) -> dict[str, Any]:
        """
        Delete a workflow.

        Args:
            workflow_id: The workflow ID

        Returns:
            Deletion result
        """
        return self._client.request("DELETE", f"/api/v1/workflows/{workflow_id}")

    def execute(
        self,
        workflow_id: str,
        inputs: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Execute a workflow.

        Args:
            workflow_id: The workflow ID
            inputs: Input data for the workflow

        Returns:
            Execution result with execution_id
        """
        return self._client.request(
            "POST",
            f"/api/v1/workflows/{workflow_id}/execute",
            json={"inputs": inputs or {}},
        )

    def get_execution(self, execution_id: str) -> dict[str, Any]:
        """
        Get workflow execution status.

        Args:
            execution_id: The execution ID

        Returns:
            Execution details
        """
        return self._client.request("GET", f"/api/v1/workflows/executions/{execution_id}")

    def list_executions(
        self,
        workflow_id: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List workflow executions.

        Args:
            workflow_id: Optional filter by workflow ID
            limit: Maximum number of executions to return
            offset: Number of executions to skip

        Returns:
            List of executions
        """
        params = {"limit": limit, "offset": offset}
        if workflow_id:
            params["workflow_id"] = workflow_id

        return self._client.request(
            "GET",
            "/api/v1/workflows/executions",
            params=params,
        )

    def cancel_execution(self, execution_id: str) -> dict[str, Any]:
        """
        Cancel a workflow execution.

        Args:
            execution_id: The execution ID

        Returns:
            Cancellation result
        """
        return self._client.request(
            "POST",
            f"/api/v1/workflows/executions/{execution_id}/cancel",
        )

    def list_templates(self) -> dict[str, Any]:
        """
        List available workflow templates.

        Returns:
            List of templates
        """
        return self._client.request("GET", "/api/v1/workflows/templates")

    def create_from_template(
        self,
        template_id: str,
        name: str,
        **overrides,
    ) -> dict[str, Any]:
        """
        Create a workflow from a template.

        Args:
            template_id: The template ID
            name: Name for the new workflow
            **overrides: Override template defaults

        Returns:
            Created workflow
        """
        return self._client.request(
            "POST",
            f"/api/v1/workflows/templates/{template_id}/create",
            json={"name": name, **overrides},
        )


class AsyncWorkflowsAPI:
    """
    Asynchronous Workflows API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     execution = await client.workflows.execute("workflow-123", {"input": "value"})
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(
        self,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List all workflows."""
        return await self._client.request(
            "GET",
            "/api/v1/workflows",
            params={"limit": limit, "offset": offset},
        )

    async def get(self, workflow_id: str) -> dict[str, Any]:
        """Get a workflow by ID."""
        return await self._client.request("GET", f"/api/v1/workflows/{workflow_id}")

    async def create(
        self,
        name: str,
        steps: list[dict[str, Any]],
        description: str | None = None,
        **kwargs,
    ) -> dict[str, Any]:
        """Create a new workflow."""
        data = {"name": name, "steps": steps, **kwargs}
        if description:
            data["description"] = description

        return await self._client.request("POST", "/api/v1/workflows", json=data)

    async def update(
        self,
        workflow_id: str,
        **updates,
    ) -> dict[str, Any]:
        """Update a workflow."""
        return await self._client.request(
            "PUT",
            f"/api/v1/workflows/{workflow_id}",
            json=updates,
        )

    async def delete(self, workflow_id: str) -> dict[str, Any]:
        """Delete a workflow."""
        return await self._client.request("DELETE", f"/api/v1/workflows/{workflow_id}")

    async def execute(
        self,
        workflow_id: str,
        inputs: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Execute a workflow."""
        return await self._client.request(
            "POST",
            f"/api/v1/workflows/{workflow_id}/execute",
            json={"inputs": inputs or {}},
        )

    async def get_execution(self, execution_id: str) -> dict[str, Any]:
        """Get workflow execution status."""
        return await self._client.request("GET", f"/api/v1/workflows/executions/{execution_id}")

    async def list_executions(
        self,
        workflow_id: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List workflow executions."""
        params = {"limit": limit, "offset": offset}
        if workflow_id:
            params["workflow_id"] = workflow_id

        return await self._client.request(
            "GET",
            "/api/v1/workflows/executions",
            params=params,
        )

    async def cancel_execution(self, execution_id: str) -> dict[str, Any]:
        """Cancel a workflow execution."""
        return await self._client.request(
            "POST",
            f"/api/v1/workflows/executions/{execution_id}/cancel",
        )

    async def list_templates(self) -> dict[str, Any]:
        """List available workflow templates."""
        return await self._client.request("GET", "/api/v1/workflows/templates")

    async def create_from_template(
        self,
        template_id: str,
        name: str,
        **overrides,
    ) -> dict[str, Any]:
        """Create a workflow from a template."""
        return await self._client.request(
            "POST",
            f"/api/v1/workflows/templates/{template_id}/create",
            json={"name": name, **overrides},
        )
